composer/checkpoint/download.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2024 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Useful functions for load checkpoints from remote object store or local disk."""
+import logging
+from typing import Optional
+
+from composer.utils import (
+    dist,
+    extract_path_from_symlink,
+    maybe_create_object_store_from_uri,
+    parse_uri,
+    retry,
+)
+
+log = logging.getLogger(__name__)
+
+
+def download_file(
+    source_uri: str,
+    destination_path: str,
+    node_ranks: Optional[list[int]] = None,
+    num_attempts: int = 5,
+):
+    """Downloads a file (object) from the specified URI to the specified directory.
+
+    Args:
+        source_uri (str): The URI to download the file from or a symlink to the URI.
+        destination_path (str): The directory to download the file to.
+        node_ranks (list[int]): The ranks of the nodes that will download the file. If None, all nodes will download the file.
+        num_attempts (int): Retry for object store downloads. Default to 5.
+    """
+    # Only local rank 0 downloads
+    local_rank = dist.get_local_rank()
+    if local_rank != 0:
+        return
+
+    node_rank = dist.get_node_rank()
+    if node_ranks is not None and node_rank not in node_ranks:
+        return
+
+    object_store = maybe_create_object_store_from_uri(source_uri)
+    _, _, source_path = parse_uri(source_uri)
+    if source_uri.endswith('.symlink'):
+        source_path = extract_path_from_symlink(source_path, object_store)
+    assert object_store is not None
+
+    @retry(num_attempts=num_attempts)
+    def _download():
+        object_store.download_object(
+            object_name=source_path,
+            filename=destination_path,
+        )
+
+    log.debug(f'Downloading {source_path} to {destination_path}')
+    _download()
+    log.debug(f'Finished downloading {source_path} to {destination_path}')
+
+
+def download_monolithic_checkpoint(
```

Comment:
should this be all local rank 0 or global rank 0?

Reply:
global rank 0, updated the PR to make it more clear. I had a discussion with @eracah  on this, we looked the current code, it works like this: for fsdp + monolithic, only global rank 0 load the state dict, then for non-model state dict, global rank0 broadcast to other ranks. For model state dict, it uses the fsdp sync_modules_states to broadcast